storage/scratchpad/src/sparse_merkle/mod.rs

```diff
@@ -155,6 +155,25 @@ pub struct SparseMerkleTree<V> {
     inner: Arc<Inner<V>>,
 }
 
+/// A type for tracking intermediate hashes at sparse merkle tree nodes in between batch
+/// updates by transactions. It contains tuple (hash_value, txn_id, single_new_leaf), where
+/// hash_value is the value after all the updates by transaction txn_id (txn_id-th batch)
+/// and single_new_leaf is a bool that's true if the node subtree contains one new leaf.
+/// (this is needed to recursively merge IntermediateHashes).
+type IntermediateHashes = Vec<(usize, HashValue, bool)>;
+
+trait KeyHashValue {
```

Comment:
this is to avoid having two different binary search partition methods w. different names. if you'd like that better happy to do it that way.

Reply:
maybe just copy this until we have 1.52 https://doc.rust-lang.org/src/core/slice/mod.rs.html#3415-3417

Reply:
good point - I used (HashValue, (usize, &V))